### DISCUSSION TITLE
How to map graphql inputs to prisma query to create filters

### DISCUSSION THREAD

**Discussion by nenadfilipovic** [prisma/prisma#7976] [Q&A]
How can I manually create filters that nexus-prisma-plugin creates automatically?

I am using graphql-nexus with new nexus-prisma plugin that does not have crud api yet,
this is input that will represent my filter on type Device

```js
DeviceWhereInput: inputObjectType({
  name: 'DeviceWhereInput',
  description: 'Device where input arguments',
  definition(t) {
    t.field(id.name, {
      type: 'StringFilter',
      description: id.description,
    });
    t.field(name.name, {
      type: 'StringFilter',
      description: name.description,
    });
    t.field(description.name, {
      type: 'StringFilter',
      description: description.description,
    });
    t.field(isPublic.name, {
      type: 'BooleanFilter',
      description: isPublic.description,
    });
    t.field(createdAt.name, {
      type: 'DateTimeFilter',
      description: createdAt.description,
    });
    t.list.nonNull.field('OR', {
      type: DeviceInput.DeviceWhereInput,
      description: 'Logical OR operator',
    });
    t.list.nonNull.field('AND', {
      type: DeviceInput.DeviceWhereInput,
      description: 'Logical AND operator',
    });
    t.list.nonNull.field('NOT', {
      type: DeviceInput.DeviceWhereInput,
      description: 'Logical NOT operator',
    });
  },
}),
```

```js
StringFilter: inputObjectType({
  name: 'StringFilter',
  description: 'Filter operations on String type',
  definition(t) {
    t.field('contains', {
      type: 'String',
      description: 'String contains',
    });
    t.field('equals', {
      type: 'String',
      description: 'String is equal to',
    });
    t.field('startsWith', {
      type: 'String',
      description: 'String starts with',
    });
    t.field('endsWith', {
      type: 'String',
      description: 'String ends with',
    });
  },
}),
```

~~First problem is types for AND, OR and NOT are not created, I cannot find them under `NexusGenInputs['UserWhereInput']` because that type is not created, but they exist in graphql.schema.~~ Fixed, it seems that input type will not be generated if that input is not used in schema.

Second problem is when I pass this as arg to resolver, I suppose whole structure of this input reflects how prisma query looks and I should just pass it inside prisma query

For this graphql query:

```gql
query {
  findOneUser(
    where: {
      id: { contains: "asd" }
      OR: { name: { equals: "sad" }, AND: { isPublic: { not: false } } }
    }
  ) {
    id
  }
}

```

Console log of this arg looks like this

```js
 {
    id: { contains: 'asd' },
    OR: [{ name: { equals: 'sad' }, AND: [{ isPublic: { not: false } }] }],
 };
```
And I can pass it like this

```js
return await prisma.user.findFirst({
  where: {
    ...where,
  },
});
```

This works but typescript is complaining a lot, I could go and check every possible field with `something != null ? something : undefined` but is there any better way, how you at Prisma solve this.

Update:

Something like this gives no errors:

```ts
return await prisma.user.findFirst({
  where: {
    ...(where as Prisma.UserWhereInput),
  },
});
```

Update 2:

Maybe can I reuse this function https://github.com/graphql-nexus/nexus-plugin-prisma/blob/6c8801c6e1d99bfdb73a7c1c89db9607712b0e01/src/null.ts

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
